Graphics_Engine/Query/Color.cs

```diff
@@ -0,0 +1,107 @@
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Geometry;
+using BH.oM.Graphics;
+using BH.oM.Reflection.Attributes;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Drawing;
+using System.Linq;
+
+namespace BH.Engine.Graphics
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /****           Public Methods                  ****/
+        /***************************************************/
+
+        [Description("Gets a color from a gradient at the spacified position")]
+        [Input("gradient", "The gradient to Query a Color from")]
+        [Input("val", "The number between 0 and 1 to use for interpolation of the markers colors")]
+        [Output("color", "The Color at the specified position")]
+        public static Color Color(this Gradient gradient, double val)
+        {
+            // keep val in the domain
+            val = val > 1 ? 1 : val < 0 ? 0 : val;
+
+            // Find adjacent markers
+            KeyValuePair<decimal, Color> upper = gradient.Markers.FirstOrDefault(x => x.Key >= (decimal)val);
+            KeyValuePair<decimal, Color> lower = gradient.Markers.LastOrDefault(x => x.Key <= (decimal)val);
+
+            // Check adjacency on both sides
+            if (lower.Equals(default(KeyValuePair<decimal, Color>)))
+                return gradient.Markers.First().Value;
+            if (upper.Equals(default(KeyValuePair<decimal, Color>)))
+                return gradient.Markers.Last().Value;
+
+            // Interpolate
+            return InterpolateColor(lower, upper, (decimal)val);
+        }
+
+        /***************************************************/
+
+        [Description("Gets a color from a gradient at the spacified position scaled between from and to")]
+        [Input("gradient", "The gradient to Query a Color from")]
+        [Input("val", "The number between 'from' and 'to' to use for interpolation of the markers colors")]
+        [Input("from", "The lower bound of 'val's rescaling to 0 to 1")]
+        [Input("to", "The upper bound of 'val's rescaling to 0 to 1")]
+        [Output("color", "The Color at the specified position")]
+        public static Color Color(this Gradient gradient, double val, double from, double to)
```

Comment:
If I read this correctly, this method assumes the gradient to have a range between 0 and 1, but we are not enforcing that anywhere for the creation of the gradients.
For this to work as expected, I think the gradient needs to be checked/we need to update the Create method for the gradient so that it always lies between 0 and 1?

Please ignore if I am missing something here.

Reply:
Markers outside of the range won't create any issues apart from being pointless, but added an warning